### PR TITLE
check omegaconf gpus

### LINKED ISSUE #2270: Exception when trainer.gpus are passed as a list through hydra (omegaconf)
## 🐛 Bug
When we pass the `gpus` flag for `Trainer` using `hydra` (which in turn uses `Omegaconf`) it is passed as `omegaconf.listconfig.ListConfig`. However, internally `Trainer` only checks for `int`, `str`, `list`. This throws an exception. 

#### Code sample
```
#!/usr/bin/env python

import warnings
warnings.filterwarnings("ignore")

import os

import torch
from torch import nn
from torch.utils.data import DataLoader, random_split
from torchvision.datasets import MNIST
from torchvision import datasets, transforms
from torch.nn import functional as F
from torch import optim

from pytorch_lightning.core.lightning import LightningModule
from pytorch_lightning import Trainer
from omegaconf import OmegaConf

class LitMNIST(LightningModule):
  def __init__(self):
      super().__init__()
      self.layer_1 = torch.nn.Linear(28 * 28, 128)
      self.layer_2 = torch.nn.Linear(128, 256)
      self.layer_3 = torch.nn.Linear(256, 10)  
      
  def prepare_data(self):
    transform=transforms.Compose([transforms.ToTensor(), transforms.Normalize((0.1307,), (0.3081,))])
    # transform
    mnist_train = MNIST(os.getcwd(), train=True, download=False, transform=transforms.ToTensor())
    
    # train/val split
    mnist_train, mnist_val = random_split(mnist_train, [55000, 5000])

    # assign to use in dataloaders
    self.train_dataset = mnist_train
    self.val_dataset = mnist_val
    
  def forward(self, x):
      batch_size, channels, width, height = x.size()
      x = x.view(batch_size, -1)
      x = self.layer_1(x)
      x = torch.relu(x)
      x = self.layer_2(x)
      x = torch.relu(x)
      x = self.layer_3(x)
      x = torch.log_softmax(x, dim=1)
      return x    
    
  def training_step(self, batch, batch_idx):
      x, y = batch
      logits = self(x)
      loss = F.nll_loss(logits, y)

      # add logging
      logs = {'loss': loss}
      return {'loss': loss, 'log': logs}    
    
  def validation_step(self, batch, batch_idx):
    x, y = batch
    logits = self(x)
    loss = F.nll_loss(logits, y)
    return {'val_loss': loss}    

  def train_dataloader(self):
    return DataLoader(self.train_dataset, batch_size=64)
  
  def val_dataloader(self):
    return DataLoader(self.val_dataset, batch_size=64) 
  
  def configure_optimizers(self):
    return optim.Adam(self.parameters(), lr=1e-3)    

if __name__=='__main__':
#   trainer_params = OmegaConf.create({'max_epochs': 1, 'overfit_pct': 0.01, 'gpus': 1}) # this works
#   trainer_params = OmegaConf.create({'max_epochs': 1, 'overfit_pct': 0.01, 'gpus': [1]}) # this throws an exception about gpus not being int, str, or list(int)
  model = LitMNIST()
  trainer = Trainer(**trainer_params)
  trainer.fit(model)
```

### PR DESCRIPTION
**Pull request by Borda**
## What does this PR do?

Fixes #2270

# Before submitting

- [ ] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you create a separate PR for every change.
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests? 
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/trainer/distrib_parts.py [hunk 1]
```diff
@@ -362,8 +362,8 @@ def check_gpus_data_type(gpus: Any) -> None:
 
 def normalize_parse_gpu_input_to_list(gpus: Union[int, List[int]]) -> Optional[List[int]]:
     assert gpus is not None
-    if isinstance(gpus, list):
-        return gpus
+    if isinstance(gpus, MutableSequence):
```
Comment (omry) [inline pytorch_lightning/trainer/distrib_parts.py:365]: lgtm, thanks!

**Comment by omry** [comment]
@Borda,
A followup can be to clean up the other special cases in the code.
With the exception of the serialization change, I think you can drop most explicit imports of OmegaConf and just test against MutableSequence and MutableMapping.

**Comment by Borda** [comment]
> With the exception of the serialization change, I think you can drop most explicit imports of OmegaConf and just test against MutableSequence and MutableMapping.

that is what I just did in the previous commit :]
cc: @williamFalcon

File: tests/trainer/test_trainer.py [hunk 2]
```diff
@@ -905,6 +906,15 @@ def __init__(self, **kwargs):
         TrainerSubclass(abcdefg='unknown_arg')
 
 
+@pytest.mark.parametrize('trainer_params', [
+    OmegaConf.create({'max_epochs': 1, 'gpus': 1}),
+    OmegaConf.create({'max_epochs': 1, 'gpus': [0]}),
+])
+@pytest.mark.skipif(not torch.cuda.is_available(), reason="test requires GPU machine")
+def test_trainer_omegaconf(trainer_params):
+    _ = Trainer(**trainer_params)
```
Comment (omry) [inline tests/trainer/test_trainer.py:915]: I personally prefer to just call the constructor in such cases (your call of course).
```suggestion
    Trainer(**trainer_params)
```

**Comment by omry** [comment]
> > With the exception of the serialization change, I think you can drop most explicit imports of OmegaConf and just test against MutableSequence and MutableMapping.
> 
> that is what I just did in the previous commit :]
> cc: @williamFalcon

Oh, missed the stack.
I reviewed the whole change.

File: pytorch_lightning/core/saving.py [hunk 3]
```diff
@@ -6,20 +6,18 @@
...
+from typing import Union, Dict, Any, Optional, Callable, MutableMapping
 
 from pytorch_lightning import _logger as log
 from pytorch_lightning.utilities import rank_zero_warn, AttributeDict
 from pytorch_lightning.utilities.cloud_io import load as pl_load
 
 PRIMITIVE_TYPES = (bool, int, float, str)
-ALLOWED_CONFIG_TYPES = (AttributeDict, dict, Namespace)
+ALLOWED_CONFIG_TYPES = (AttributeDict, MutableMapping, Namespace)
 try:
```
[the remark this answers is not in the document]
Comment (Borda) [reply pytorch_lightning/core/saving.py:17]: no, it is needed for export to YAML

File: pytorch_lightning/trainer/distrib_parts.py [hunk 4]
```diff
@@ -356,14 +356,14 @@ def check_gpus_data_type(gpus: Any) -> None:
...
+    if gpus is not None and (not isinstance(gpus, (int, str, MutableSequence)) or isinstance(gpus, bool)):
+        raise MisconfigurationException("GPUs must be int, string or sequence of ints or None.")
 
 
 def normalize_parse_gpu_input_to_list(gpus: Union[int, List[int]]) -> Optional[List[int]]:
     assert gpus is not None
-    if isinstance(gpus, list):
-        return gpus
+    if isinstance(gpus, MutableSequence):
+        return list(gpus)
```
[the remark this answers is not in the document]
Comment (Borda) [reply pytorch_lightning/trainer/distrib_parts.py:366]: rather have everywhere the same typo

File: pytorch_lightning/core/saving.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (omry) [reply pytorch_lightning/core/saving.py:17]: Gotcha.
didn't see it in this file.

I recommend:

A conditional:
```
from omegaconf import OmegaConf
```

(Container is kind of an implementation detail).

When you want to check if an object is an OmegaConf config, use one of [these](https://omegaconf.readthedocs.io/en/latest/usage.html#omegaconf-is-config-is-dict-is-list).

(Specifically here I would go with OmegaConf.is_config()).

**Comment by williamFalcon** [comment]
glad we could get this to work!

**Pull request merged**